### PR TITLE
Directory Listing & Web Scrapping Links

### PR DESCRIPTION
**Pull request by godylockz**
Finally a directory fuzzer that handles directory listings AND web scrapping links.

## Branching checklist
- [Yes] There is an issue associated with your PR (bug, feature, etc.. if not, create one)
- [Yes ] Your PR description references the associated issue (i.e. fixes #123456)
https://github.com/epi052/feroxbuster/issues/291
https://github.com/epi052/feroxbuster/issues/245
- [Yes] Code is in its own branch
- [Yes ] Branch name is related to the PR contents
- [Yes] PR targets main

## Static analysis checks
- [Yes] All rust files are formatted using `cargo fmt`
- [Yes] All `clippy` checks pass when running `cargo clippy --all-targets --all-features -- -D warnings -A clippy::mutex-atomic`
- [Yes] All existing tests pass

## Documentation
- [ Yes] New code is documented using [doc comments](https://doc.rust-lang.org/stable/rust-by-example/meta/doc.html)
- [N/A] Documentation about your PR is included in the README, as needed

## Additional Tests
- [Added 2 tests] New code is unit tested
- [Added 2 tests] New code is integration tested, as needed
- [Yes] New tests pass

## Notes
For best performance on directory listings, I recommend Redirections and Extract Links are on
2 existing extraction tests required changes to account for the single additional request to directories to check for directory listings

### CODE REVIEW

**Review by epi052 — CHANGES_REQUESTED** [review]
mostly done, there's a few things i want to look at still, but pushing what requests i have now so you have them sooner rather than later

File: src/extractor/container.rs [hunk 1]
```diff
@@ -370,18 +390,91 @@ impl<'a> Extractor<'a> {
...
+        log::trace!("enter: get_links_by_attr");
+
+        let selector = Selector::parse(html_tag).unwrap();
+        let tags = html
+            .select(&selector)
+            .filter(|a| a.value().attrs().any(|attr| attr.0 == html_attr));
+        for t in tags {
+            if let Some(link) = t.value().attr(html_attr) {
+                log::debug!("Parsed link \"{}\" from {}", link, resp_url);
+                let mut new_url = Url::parse(&self.url).unwrap();
```
Comment (epi052) [inline src/extractor/container.rs:461]: unwrap here leads to runtime problems `thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: RelativeUrlWithoutBase', src/extractor/container.rs:461:57`

File: src/extractor/container.rs [hunk 2]
```diff
@@ -370,18 +390,91 @@ impl<'a> Extractor<'a> {
...
+                new_url.set_path(link);
+                if self.add_all_sub_paths(new_url.path(), links).is_err() {
+                    log::warn!("could not add sub-paths from {} to {:?}", new_url, links);
+                }
+            }
+        }
```
Comment (epi052) [inline src/extractor/container.rs:467]:
```suggestion
        html.select(&selector)
            .filter(|a| a.value().attrs().any(|attr| attr.0 == html_attr))
            .for_each(|tag| {
                if let Some(link) = tag.value().attr(html_attr) {
                    log::debug!("Parsed link \"{}\" from {}", link, resp_url);

                    let new_url = match Url::parse(&self.url) {
                        Ok(mut absolute) => {
                            absolute.set_path(link);
                            absolute.path().to_owned()
                        } // absolute url
                        Err(_) => link.into(), // relative
                    };

                    if self.add_all_sub_paths(&new_url, links).is_err() {
                        log::warn!("could not add sub-paths from {} to {:?}", new_url, links);
                    }
                }
            });
```

File: Cargo.toml [hunk 3]
```diff
@@ -23,6 +23,7 @@ lazy_static = "1.4"
 dirs = "4.0"
 
 [dependencies]
+select = "*"
```
[the remark this answers is not in the document]
Comment (godylockz) [reply Cargo.toml]: Fixed, using scraper!

File: src/extractor/container.rs [hunk 4]
```diff
@@ -369,19 +385,136 @@ impl<'a> Extractor<'a> {
...
+                let msg = format!(
+                    "{} {:>8} {:>8}l {:>8}w {:>8}c {} => Directory listing\n",
+                    status_colorizer(response.status().as_str()),
+                    response.method().as_str(),
+                    response.line_count().to_string(),
+                    response.word_count().to_string(),
+                    response.content_length().to_string(),
+                    response.url(),
+                );
+                ferox_print(&msg, &PROGRESS_PRINTER);
```
Comment (epi052) [inline src/extractor/container.rs:431]: ```suggestion```

File: src/scanner/ferox_scanner.rs [hunk 5]
```diff
@@ -106,6 +117,20 @@ impl FeroxScanner {
...
+        // Directory listing heuristic detection to not continue scanning
+        if dirlist_flag {
+            log::trace!("exit: scan_url -> Directory listing heuristic");
+
+            self.handles.stats.send(AddToF64Field(
+                DirScanTimes,
+                scan_timer.elapsed().as_secs_f64(),
+            ))?;
+
+            ferox_scan.finish()?;
```
Comment (epi052) [inline src/scanner/ferox_scanner.rs:129]:
```suggestion
            self.handles.stats.send(SubtractFromUsizeField(
                TotalExpected,
                progress_bar.length() as usize,
            ))?;

            progress_bar.finish_with_message(&format!("=> {}", style("Directory listing").green()));

            ferox_scan.finish()?;
```

File: src/extractor/container.rs [hunk 6]
```diff
@@ -7,17 +7,19 @@ use crate::{
...
         StatField::{LinksExtracted, TotalExpected},
     },
     url::FeroxUrl,
-    utils::{logged_request, make_request},
+    utils::{ferox_print, logged_request, make_request, status_colorizer},
     DEFAULT_METHOD,
 };
 use anyhow::{bail, Context, Result};
 use reqwest::{StatusCode, Url};
+use scraper::{Html, Selector};
```
Comment (epi052) [inline src/extractor/container.rs:22]:
```suggestion
use scraper::{Html, Selector};
use console::style;
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -166,7 +166,7 @@ primarily related to continuous integration and release deployment.
 
 feroxbuster uses the [`clippy`](https://rust-lang.github.io/rust-clippy/) code linter.
 
-The command that will ultimately be used in the CI pipeline for linting is `cargo clippy --all-targets --all-features -- -D warnings -A clippy::unnecessary_unwrap`.  
+The command that will ultimately be used in the CI pipeline for linting is `cargo clippy --all-targets --all-features -- -D warnings -A clippy::mutex-atomic -A clippy::unnecessary_unwrap`.
```
Comment (epi052) [inline CONTRIBUTING.md:169]:
```suggestion
The command that will ultimately be used in the CI pipeline for linting is `cargo clippy --all-targets --all-features -- -D warnings -A clippy::mutex-atomic`.
```

File: .github/pull_request_template.md [hunk 8]
```diff
@@ -11,7 +11,7 @@ Long form explanations of most of the items below can be found in the [CONTRIBUT
 
 ## Static analysis checks
 - [ ] All rust files are formatted using `cargo fmt`
-- [ ] All `clippy` checks pass when running `cargo clippy --all-targets --all-features -- -D warnings -A clippy::mutex-atomic`
+- [ ] All `clippy` checks pass when running `cargo clippy --all-targets --all-features -- -D warnings -A clippy::mutex-atomic -A clippy::unnecessary_unwrap`
```
Comment (epi052) [inline .github/pull_request_template.md:14]:
```suggestion
- [ ] All `clippy` checks pass when running `cargo clippy --all-targets --all-features -- -D warnings -A clippy::mutex-atomic`
```

**Review by epi052 — APPROVED** [review]

**Comment by epi052** [comment]
great job with this, thank you! merging :fire:

**Comment by SecT0uch** [comment]
Would it be possible (and wise) not to add the directory to the queue if the listing is empty ?

**Comment by godylockz** [comment]
> Would it be possible (and wise) not to add the directory to the queue if the listing is empty ?

@SecT0uch Do you have a target that has an example of this that your willing to share? 
If it is deemed repeatable, it would be an easy fix. In the meantime, just cancel the scan on that directory if you manually verified there is not any files within.
This is a "new" feature so there may potentially be bugs. This is something that isn't "broken" per-se. No other directory fuzzing tool will even care that its a directory listing, so at least there's that.
@epi052 I'm actually surprised this occurs?

**Comment by SecT0uch** [comment]
That would be reproducible like that:
```
mkdir /tmp/empty
cd /tmp/empty
python -m "http.server"
```
You get an empty listing on http://localhost:8000

**Pull request merged**